### DISCUSSION TITLE
`customize-set-variable` just about everywhere, or leave some `setq`s alone?

### DISCUSSION THREAD

**Discussion by jf** [SystemCrafters/crafted-emacs#420] [Q&A]
I previously asked about the rationale for recommending and preferring `customize-set-variable` over `setq`. It makes sense, but I'm just noticing as I look through code that I see some `setq`s already that could possibly be changed. My question is: should this preference be universal for all customizable variables? or are there situations when it's ok to just use `setq`?

For example, I see in https://github.com/SystemCrafters/crafted-emacs/blob/master/examples/straight-el/early-init.el a bunch of `setq`s (and no `customize-set-variable`s) for these variables. The first one is customizable; I'm not sure about the rest (I'm not able at this point in time to grok the crafted-emacs code to determine if `customize-set-variable` is really necessary for `crafted-package-*`):
- package-enable-at-startup
- crafted-package-system
- crafted-package-installer
- crafted-package-installed-predicate

**Comment by jf** [comment]
Just to add one more example: `custom-file` is a customizable variable as well, but the Getting Started Guide uses `setq` to set its value (see https://github.com/SystemCrafters/crafted-emacs/blob/master/docs/getting-started-guide.org#emacs-initialization)

**Comment by jvdydev** [comment]
Just looked at it, seems like `package-enable-at-startup` should, in fact, be using `customize-set-variable` (as it's declared `defcustom` in `package.el`).

As for the `crafted-package-*`, both `installer` and `installed-predicate` are declared via `defvar`, therefor should be set with `setq`.

The `crafted-package-system` is (I think) a leftover from either crafted-emacs v1 or an older version of the package-config which would bootstrap alternative package managers (tl;dr: it should be removed from the examples as it's not declared at all).

While the general goal is to use `customize-set-variable` with `defcustom` and `setq` with `defvar`, I think we have some code which doesn't conform to this rule yet, thank you for pointing them out :smile:

**Reply by sthesing** [reply]
You beat me to it, @jvdydev :smile:

**Reply by jf** [reply]
thank you for looking into the details! so to clarify: you're also saying that `setq` withOUT a `defvar` first does not work? It's not immediately obvious from the help for `setq`.

**Reply by jvdydev** [reply]
Oh sorry, I phrased that poorly.
`setq` without `defvar` *does* set the variable, it also declares it if not previously declared.
However, it is usually considered to be "bad style" to not declare variables using either `defvar` or `defcustom`.
The variable I was referring to (`crafted-package-system`) seems to be completely unused in the `crafted-emacs` code-base (including not being declared), hence I was suggesting to remove it from the examples.

Hope that makes sense :smile:

**Comment by sthesing** [comment]
I think I'm oversimplifying this a bit, but essentially

- `customize-set-variable` is to set variables defined by `defcustom`,
- `setq` is to set variables defined by `defvar`.

We could certainly look over our use of customization variables and streamline this a bit. It is inconsistent at places.
